agent/dockerclient/dockerapi/docker_events_buffer.go

```diff
@@ -51,15 +52,21 @@ func NewInfiniteBuffer() *InfiniteBuffer {
 }
 
 // StartListening starts reading from the input channel and writes to the buffer
-// TODO: wire in ctx to stop listening
-func (buffer *InfiniteBuffer) StartListening(events chan *docker.APIEvents) {
-	for event := range events {
-		go buffer.CopyEvents(event)
+// When context is cancelled, stop listening
+func (buffer *InfiniteBuffer) StartListening(ctx context.Context, eventChan <-chan events.Message) {
+	for {
+		select {
+		// If context is cancelled, return
+		case <-ctx.Done():
```

Comment:
What if the context is canceled but there is still event in the `eventChan` channel? We might miss some events in this case.

Reply:
I think Katy was implementing the "TODO: wire in ctx to stop listening". We stopped listening for events when context was cancelled: https://github.com/aws/amazon-ecs-agent/blob/dev/agent/dockerclient/dockerapi/docker_client.go#L787 
Once context is cancelled, I don't think we are interested in the events for it any more?

Reply:
Added logic to drain the channel into buffer when the context is closed. Was worried about potentially introducing inherently racy condition but the eventChan is not shared across multiple contexts so I think it's fine.